src/cabal-details.js

```diff
@@ -0,0 +1,365 @@
+        this.user.local = true
+        this.user.online = true
+        this._emitUpdate()
+      })
+    })
+  }
+
+  _initialize() {
+    const cabal = this._cabal
+    // populate channels
+    cabal.channels.get((err, channels) => {
+      channels.forEach((channel) => {
+        let details = this.channels[channel]
+        if (!details) {
+          this.channels[channel] = new ChannelDetails(cabal, channel)
+        }
+        // listen for updates that happen within the channel
+        cabal.messages.events.on(channel, this.messageListener.bind(this))
+
+        // add all users joined to a channel
+        cabal.memberships.getUsers(channel, (err, users) => {
+          users.forEach((u) => this.channels[channel].addMember(u))
+        })
+
+        // for each channel, get the topic
+        cabal.topics.get(channel, (err, topic) => {
+          this.channels[channel].topic = topic || ''
+        })
+      })
+    })
+
+    cabal.getLocalKey((err, lkey) => {
+      cabal.memberships.getMemberships(lkey, (err, channels) => {
+        if (channels.length === 0) {
+          // make `default` the first channel if no saved state exists
+          this.joinChannel('default')
+        }
+        for (let channel of channels) { 
+          // it's possible to be joined to a channel that `cabal.channels.get` doesn't return
+          // (it's an empty channel, with no messages)
+          let details = this.channels[channel]
+          if (!details) {
+            this.channels[channel] = new ChannelDetails(cabal, channel)
+            // listen for updates that happen within the channel
+            cabal.messages.events.on(channel, this.messageListener.bind(this))
+          }
+          this.channels[channel].joined = true
+        }
+      this._emitUpdate()
+      })
+    })
+
+    // notify when a user has joined a channel
+    this.registerListener(cabal.memberships.events, 'add', (channel, user) => {
+      if (!this.channels[channel]) { 
+        this.channels[channel] = new ChannelDetails(this._cabal, channel)
+      }
+      this.channels[channel].addMember(user)
+      this._emitUpdate()
+    })
+
+    // notify when a user has left a channel
+    this.registerListener(cabal.memberships.events, 'remove', (channel, user) => {
+      if (!this.channels[channel]) { 
+        this.channels[channel] = new ChannelDetails(this._cabal, channel)
+      }
+      this.channels[channel].removeMember(user)
+      this._emitUpdate()
+    })
+
+    // register to be notified of new channels as they are created
+    this.registerListener(cabal.channels.events, 'add', (channel) => {
+      let details = this.channels[channel]
+      if (!details) {
+        this.channels[channel] = new ChannelDetails(cabal, channel)
+      }
+      // TODO: only do this for our joined channels, instead of all channels
+      // Calls fn with every new message that arrives in channel.
+      cabal.messages.events.on(channel, this.messageListener.bind(this))
```

Comment:
Like the others above: needs to be unregistered on instance destruction.

Reply:
moving this up into future pr, tracked here https://github.com/cabal-club/cabal-client/issues/12